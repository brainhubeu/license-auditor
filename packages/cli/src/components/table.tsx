import { Box, Static, Text } from "ink";
import { useMemo, type ReactNode } from "react";

export interface Column<T extends Record<string, string>> {
  title: string;
  accessor: keyof T;
  cell?: (rowData: T) => ReactNode;
}

interface TableProps<T extends Record<string, string>> {
  columns: Column<T>[];
  data: T[];
}

export function Table<T extends Record<string, string>>({
  columns,
  data,
}: TableProps<T>) {
  const columnsWithWidth = useMemo(() => {
    const columnsWithWidth = [];
    for (const column of columns) {
      const columnWidth = Math.max(
        ...data.map((row) => (row[column.accessor] ?? "").length),
        column.title.length,
      );
      columnsWithWidth.push({ ...column, width: columnWidth });
    }
    return columnsWithWidth;
  }, [columns, data]);

  const createHorizontalLine = (
    leftChar: string,
    midChar: string,
    rightChar: string,
  ) => {
    let line = leftChar;
    for (let i = 0; i < columnsWithWidth.length; i++) {
      line += "─".repeat((columnsWithWidth[i]?.width ?? 0) + 2); // +2 for padding
      if (i < columns.length - 1) {
        line += midChar;
      }
    }
    line += rightChar;
    return line;
  };

  const topLine = createHorizontalLine("┌", "┬", "┐");
  const headerSeparator = createHorizontalLine("├", "┼", "┤");
  const bottomLine = createHorizontalLine("└", "┴", "┘");

  const headerRow = (
    <Box flexDirection="row" key="header-row">
      <Text>│</Text>
      {columnsWithWidth.map((columnWithWidth) => {
        const content = columnWithWidth.title.padEnd(
          columnWithWidth.width,
          " ",
        );
        return (
          <Text key={columnWithWidth.title}>
            {" "}
            {content} {"│"}
          </Text>
        );
      })}
    </Box>
  );

  const tableContent = [
    <Text key="top-line">{topLine}</Text>,
    headerRow,
    <Text key="header-separator">{headerSeparator}</Text>,
    ...data.map((row, rowIndex) => (
      <Box key={Object.values(row).join(",")} flexDirection="row">
        <Text>│</Text>
        {columnsWithWidth.map((columnWithWidth, colIndex) => {
          const cellValue = row[columnWithWidth.accessor]?.toString() ?? "";
          const cellContent = columnWithWidth.cell ? (
            columnWithWidth.cell(row)
          ) : (
            <Text>{cellValue.padEnd(columnWithWidth.width, " ")}</Text>
          );
          return (
            <Text key={`${columnWithWidth.title}-${rowIndex}-${colIndex}`}>
              {" "}
              {cellContent} {"│"}
            </Text>
          );
        })}
      </Box>
    )),
    <Text key="bottom-line">{bottomLine}</Text>,
  ];

  // render entire table content at once with Static to prevent re-renders
  return <Static items={tableContent}>{(item) => item}</Static>;
}
