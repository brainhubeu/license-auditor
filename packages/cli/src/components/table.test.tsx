import { Text } from "ink";
import { render } from "ink-testing-library";
import { describe, expect, it, vi } from "vitest";
import { type Column, Table } from "./table.js";

type RowData = {
  name: string;
  description: string;
  age: string;
};

vi.mock("../hooks/use-terminal-dimensions.js", async () => {
  const actual = await vi.importActual("../hooks/use-terminal-dimensions.js");
  return {
    ...actual,
    useTerminalDimensions: vi.fn(),
  };
});

describe("Table Component with mocked terminal dimensions", () => {
  const columns: Column<RowData>[] = [
    { title: "Name", accessor: "name" },
    {
      title: "Description",
      accessor: "description",
      cell: (content) => <Text color="red">{content}</Text>,
    },
    { title: "Age", accessor: "age" },
  ];

  const data: RowData[] = [
    {
      name: "Alice",
      description:
        "Develops and maintains software, writes code, tests new features, and fixes bugs.",
      age: "20",
    },
    {
      name: "Bob",
      description:
        "Creates robust applications by programming, testing, and deploying software solutions.",
      age: "30",
    },
    {
      name: "Charlie",
      description:
        "Responsible for designing and implementing user interfaces, ensuring a seamless user experience.",
      age: "27",
    },
  ];

  it("renders the table with sufficient terminal width", async () => {
    const { useTerminalDimensions } = await import(
      "../hooks/use-terminal-dimensions.js"
    );
    // @ts-expect-error
    useTerminalDimensions.mockReturnValue([350, 20]);

    const { lastFrame } = render(<Table columns={columns} data={data} />);
    const output = lastFrame();

    expect(output).toMatchSnapshot();
  });

  it("renders the table with insufficient terminal width", async () => {
    const { useTerminalDimensions } = await import(
      "../hooks/use-terminal-dimensions.js"
    );
    // @ts-expect-error
    useTerminalDimensions.mockReturnValue([50, 20]);

    const { lastFrame } = render(<Table columns={columns} data={data} />);
    const output = lastFrame();

    expect(output).toMatchSnapshot();
  });
});
