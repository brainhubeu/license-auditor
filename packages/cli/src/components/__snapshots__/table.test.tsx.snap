// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table Component with mocked terminal dimensions > renders the table with insufficient terminal width 1`] = `
"┌─────────┬────────────────────────────────┬─────┐
│ Name    │ Description                    │ Age │
├─────────┼────────────────────────────────┼─────┤
│ Alice   │ Develops and maintains         │ 20  │
│         │ software, writes code, tests   │     │
│         │ new features, and fixes bugs.  │     │
│ Bob     │ Creates robust applications by │ 30  │
│         │ programming, testing, and      │     │
│         │ deploying software solutions.  │     │
│ Charlie │ Responsible for designing and  │ 27  │
│         │ implementing user interfaces,  │     │
│         │ ensuring a seamless user       │     │
│         │ experience.                    │     │
└─────────┴────────────────────────────────┴─────┘
"
`;

exports[`Table Component with mocked terminal dimensions > renders the table with sufficient terminal width 1`] = `
"┌─────────┬──────────────────────────────────────────────────────────────────────────────────────────────────┬─────┐
│ Name    │ Description                                                                                      │ Age │
├─────────┼──────────────────────────────────────────────────────────────────────────────────────────────────┼─────┤
│ Alice   │ Develops and maintains software, writes code, tests new features, and fixes bugs.                │ 20  │
│ Bob     │ Creates robust applications by programming, testing, and deploying software solutions.           │ 30  │
│ Charlie │ Responsible for designing and implementing user interfaces, ensuring a seamless user experience. │ 27  │
└─────────┴──────────────────────────────────────────────────────────────────────────────────────────────────┴─────┘



"
`;
